joplin/importer/page_importer.py

```diff
@@ -0,0 +1,86 @@
+from urllib.parse import urlparse, parse_qs
+from pathlib import Path
+from gql import gql, Client
+from gql.transport.requests import RequestsHTTPTransport
+import json
+from importer.queries import queries
+
+from pages.topic_collection_page.factories import create_topic_collection_page_from_page_dictionary
+from pages.topic_page.factories import create_topic_page_from_page_dictionary
+from pages.information_page.factories import create_information_page_from_page_dictionary
+
+ENDPOINTS = {
+    'janis.austintexas.io': 'https://joplin-staging.herokuapp.com/api/graphql'
+}
+
+
+class PageImporter:
+
+    # todo maybe this should be static so we don't need to make an importer instance
+    def create_page(self, page_dictionary, revision_id):
+        page_creator_dictionary = {
+            'topiccollection': create_topic_collection_page_from_page_dictionary,
+            'topic': create_topic_page_from_page_dictionary,
+            'information': create_information_page_from_page_dictionary
+        }
+
+        page = page_creator_dictionary[self.page_type](page_dictionary, revision_id)
+
+        return page
+
+    def get_page_dictionary_from_revision(self):
+        sample_transport = RequestsHTTPTransport(
+            url=self.joplin_api_endpoint,
+            # use_json=True,
+            # headers={
+            #     "Content-type": "application/json",
+            # },
+            verify=False
+        )
+
+        client = Client(
+            retries=3,
+            transport=sample_transport,
+            fetch_schema_from_transport=True,
+        )
+
+        result = client.execute(queries[self.page_type], variable_values=json.dumps({'id': self.revision_id}))
+        revision_node = result['allPageRevisions']['edges'][0]['node']
+
+        # this gets us into the 'as____Page' stuff
+        page_dictionary_from_revision = next(iter(revision_node.values()))
+        return page_dictionary_from_revision
+
+    def parse_janis_preview_url(self, path):
+        # for now, just assuming these parts are always right
+        self.language = path.parts[1]
+        self.page_type = path.parts[3]
+        self.revision_id = path.parts[4]
+
+    def parse_url(self):
+        # get a urllib.parse result to play with
+        parse_result = urlparse(self.url_to_parse)
+
+        # get a joplin api endpoint
+        if parse_result.hostname in ENDPOINTS:
+            self.joplin_api_endpoint = ENDPOINTS[parse_result.hostname]
+
+        if 'CMS_API' in parse_result.query:
+            qs = parse_qs(parse_result.query)
+            self.joplin_api_endpoint = qs['CMS_API'][0]
+
+        # get a path object to play with
+        path = Path(parse_result.path)
+
+        # figure out if we're a janis preview url
+        if 'preview' in path.parts:
+            self.parse_janis_preview_url(path)
+
+    def __init__(self, url):
```

Comment:
More documentation here would help. PageImporter takes a preview_url, and parses it in order to get the information we need in order to make a page. I think it would also read clearer if parse_url wasn't a separate method and the logic was called directly in __init__.

Reply:
could probably get a few vars off the class that way too